### PR TITLE
sql: migrate has_schema_privilege from evalPrivilegeCheck to ctx.Plan…

### PR DESCRIPTION
**Pull request by ecwall**
…ner.HasPrivilege

refs https://github.com/cockroachdb/cockroach/issues/66173

Migrate has_schema_privilege from evalPrivilegeCheck to ctx.Planner.HasPrivilege.

Release note: None

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/privilege_builtins [hunk 1]
```diff
@@ -603,21 +603,21 @@ query BB
...
+false  false
 
 query BB
 SELECT has_schema_privilege((SELECT oid FROM pg_namespace WHERE nspname = 'public'), 'CREATE'),
        has_schema_privilege((SELECT oid FROM pg_namespace WHERE nspname = 'public'), 'USAGE')
 ----
 true  true
 
-query error pgcode 3F000 schema 'does_not_exist' does not exist
+query error pgcode 3F000 unknown schema "does_not_exist"
```
Comment (ecwall) [inline pkg/sql/logictest/testdata/logic_test/privilege_builtins:620]: This is controlled by an error shared with other code, but it should probably be changed for consistency.

File: pkg/sql/logictest/testdata/logic_test/privilege_builtins [hunk 2]
```diff
@@ -603,21 +603,21 @@ query BB
 SELECT has_schema_privilege((SELECT oid FROM pg_namespace WHERE nspname = 'crdb_internal'), 'CREATE'),
        has_schema_privilege((SELECT oid FROM pg_namespace WHERE nspname = 'crdb_internal'), 'USAGE')
 ----
-true  true
+false  false
```
Comment (ecwall) [inline pkg/sql/logictest/testdata/logic_test/privilege_builtins:606]: @otan @rafiss @RichardJCai - Do all of these test changes make sense?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-MtkLvcHubrUEdLXI1VK-r1-606:-Mto1N8i0vaUrukWHpCZ:bv1063f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L606)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

@otan @rafiss @RichardJCai - Do all of these test changes make sense?
</blockquote></details>

i think we want `USAGE` to still be true for all of the virtual schemas. to do that, we should update this line: https://github.com/cockroachdb/cockroach/blob/eb26eb323b14c2e00d2bf02bbd02cc7f84ef2a93/pkg/sql/catalog/schemadesc/virtual_schema_desc.go#L58

instead of giving `SELECT` to `public` it should give `USAGE`.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1638 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mto2loyB5m9aF9lgPOz:-Mto2loyB5m9aF9lgPP-:bc79l8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/sem/builtins/pg_builtins.go#L1638)):*
> ```Go
> 			}
> 			if len(databaseName) == 0 {
> 				// If no database is set, return NULL.
> ```

why do we need this check on the databaseName? should it happen before the call to `schemaHasPrivilegeSpecifier`?

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1638 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mto2loyB5m9aF9lgPOz:-MtoIq0o8jrvOIIR9et3:btmpts1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/sem/builtins/pg_builtins.go#L1638)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why do we need this check on the databaseName? should it happen before the call to `schemaHasPrivilegeSpecifier`?
</blockquote></details>

I'm not sure why it happens in this order, but the return is the same place as it was originally. 
I just combined the 2 ifs into 1.
```
			if len(ctx.SessionData().Database) == 0 {
				// If no database is set, return NULL.
				retNull = true
			}

			privs, err := parsePrivilegeStr(args[1], privMap{
				"CREATE":                   {Kind: privilege.CREATE},
				"CREATE WITH GRANT OPTION": {Kind: privilege.CREATE, GrantOption: true},
				"USAGE":                    {Kind: privilege.USAGE},
				"USAGE WITH GRANT OPTION":  {Kind: privilege.USAGE, GrantOption: true},
			})
			if err != nil {
				return nil, err
			}
			if retNull {
				return tree.DNull, nil
			}
```
___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-MtkLvcHubrUEdLXI1VK-r1-606:-MtoMMrHBdAdv5b10WHV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L606)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we want `USAGE` to still be true for all of the virtual schemas. to do that, we should update this line: https://github.com/cockroachdb/cockroach/blob/eb26eb323b14c2e00d2bf02bbd02cc7f84ef2a93/pkg/sql/catalog/schemadesc/virtual_schema_desc.go#L58

instead of giving `SELECT` to `public` it should give `USAGE`.
</blockquote></details>

Done.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-MtkLvcHubrUEdLXI1VK-r1-606:-Mu2TXdt4FAVISjVlvBa:bj6vpzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L606)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

this change lgtm now
___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-MtkLvcHubrUEdLXI1VK-r1-620:-Mu2TZQy0fxpysZ7LSSQ:b-3mn5wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb03ff285bf54189fc9499e48c2a10c53352e37/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L620)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

This is controlled by an error shared with other code, but it should probably be changed for consistency.
</blockquote></details>

i think we can address this later
___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2TAql9DUbyr4_eWyJ:-Mu2TAql9DUbyr4_eWyK:b-q2wszl) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1311)):*
> ```
> SELECT has_schema_privilege('testuser', 's', 'usage')
> ----
> true
> ```

hm, i'm not sure why this test ended up needing to change. i don't think it should have, since `s` is not a virtual schema. do you have any idea why? (as per the above line, this priv should have just been revoked)
___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2T8035umxlqII3Zae:-Mu2T8035umxlqII3Zaf:b-15zu1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1340)):*
> ```
>        has_schema_privilege('testuser2', 's', 'usage')
> ----
> true true true true
> ```

i also don't this this one should have changed

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2TAql9DUbyr4_eWyJ:-MuBopLo2ZAK9VJZVc3C:bges2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1311)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, i'm not sure why this test ended up needing to change. i don't think it should have, since `s` is not a virtual schema. do you have any idea why? (as per the above line, this priv should have just been revoked)
</blockquote></details>

It looks like `testuser` is the owner
```
user testuser

statement ok
use my_db

statement ok
CREATE SCHEMA s;
```
Which is checked here
```
func (p *planner) CheckPrivilegeForUser(
	ctx context.Context,
	descriptor catalog.Descriptor,
	privilege privilege.Kind,
	user security.SQLUsername,
) error {
	// Verify that the txn is valid in any case, so that
	// we don't get the risk to say "OK" to root requests
	// with an invalid API usage.
	if p.txn == nil || !p.txn.IsOpen() {
		return errors.AssertionFailedf("cannot use CheckPrivilege without a txn")
	}

	// Test whether the object is being audited, and if so, record an
	// audit event. We place this check here to increase the likelihood
	// it will not be forgotten if features are added that access
	// descriptors (since every use of descriptors presumably need a
	// permission check).
	p.maybeAudit(descriptor, privilege)

	privs := descriptor.GetPrivileges()

	// Check if the 'public' pseudo-role has privileges.
	if privs.CheckPrivilege(security.PublicRoleName(), privilege) {
		return nil
	}

	hasPriv, err := p.checkRolePredicate(ctx, user, func(role security.SQLUsername) bool {
		return IsOwner(descriptor, role) || privs.CheckPrivilege(role, privilege)
	})
	if err != nil {
		return err
	}
	if hasPriv {
		return nil
	}
	return pgerror.Newf(pgcode.InsufficientPrivilege,
		"user %s does not have %s privilege on %s %s",
		user, privilege, descriptor.DescriptorType(), descriptor.GetName())
}
```



_Code quote:_
```
query B
SELECT has_schema_privilege('testuser', 's', 'usage')
----
true
```

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2TAql9DUbyr4_eWyJ:-MuBphabCgCK3Z5QlW6k:b-ykwkar) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1311)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It looks like `testuser` is the owner
```
user testuser

statement ok
use my_db

statement ok
CREATE SCHEMA s;
```
Which is checked here
```
func (p *planner) CheckPrivilegeForUser(
	ctx context.Context,
	descriptor catalog.Descriptor,
	privilege privilege.Kind,
	user security.SQLUsername,
) error {
	// Verify that the txn is valid in any case, so that
	// we don't get the risk to say "OK" to root requests
	// with an invalid API usage.
	if p.txn == nil || !p.txn.IsOpen() {
		return errors.AssertionFailedf("cannot use CheckPrivilege without a txn")
	}

	// Test whether the object is being audited, and if so, record an
	// audit event. We place this check here to increase the likelihood
	// it will not be forgotten if features are added that access
	// descriptors (since every use of descriptors presumably need a
	// permission check).
	p.maybeAudit(descriptor, privilege)

	privs := descriptor.GetPrivileges()

	// Check if the 'public' pseudo-role has privileges.
	if privs.CheckPrivilege(security.PublicRoleName(), privilege) {
		return nil
	}

	hasPriv, err := p.checkRolePredicate(ctx, user, func(role security.SQLUsername) bool {
		return IsOwner(descriptor, role) || privs.CheckPrivilege(role, privilege)
	})
	if err != nil {
		return err
	}
	if hasPriv {
		return nil
	}
	return pgerror.Newf(pgcode.InsufficientPrivilege,
		"user %s does not have %s privilege on %s %s",
		user, privilege, descriptor.DescriptorType(), descriptor.GetName())
}
```
</blockquote></details>

thanks for tracking it down! that's also correct then, since owners should have all privileges.

would be worth adding another test case (if one doesn't exist already), to make sure a non-owner doesn't have USAGE, and that USAGE can be revoked from non-owners

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2T8035umxlqII3Zae:-MuBq7sUBK6PBcQaOqCO:b-eknpsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1340)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i also don't this this one should have changed
</blockquote></details>

Same as above regarding the ower.

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2TAql9DUbyr4_eWyJ:-MuHNVaD2Yzh7fZ5dnQ6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1311)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thanks for tracking it down! that's also correct then, since owners should have all privileges.

would be worth adding another test case (if one doesn't exist already), to make sure a non-owner doesn't have USAGE, and that USAGE can be revoked from non-owners
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-MuHNWZL8TRZ3PIZ6J06:-MuHNWZL8TRZ3PIZ6J07:boe06k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c96bccf9432a5037ac5751c47e019486095c075/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1349)):*
> ```
> REVOKE testuser FROM testuser2
> 
> # Schema owner has all privileges
> ```

I changed the 2 existing tests to not take the owner into account and added a separate test for checking the schema owner's privileges.

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/privilege_builtins, line 1311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-Mu2TAql9DUbyr4_eWyJ:-MuHOXW5ElT_qkripvXN:bnwpxss) ([raw file](https://github.com/cockroachdb/cockroach/blob/137d78b4d58f7cef3f2cc261e9c8cf01b7787d8a/pkg/sql/logictest/testdata/logic_test/privilege_builtins#L1311)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

revoke usage is already being tested here
```
# Privilege check on direct member of role with CREATE privilege granted.
query B
SELECT has_schema_privilege('testuser', 's', 'create')
----
true

statement ok
REVOKE USAGE ON SCHEMA s FROM testuser;

# Privilege check on direct member of role without USAGE privilege granted.
query B
SELECT has_schema_privilege('testuser', 's', 'usage')
----
false
```

**Review by rafiss — APPROVED** [review]
lgtm!

> Reviewed 2 of 6 files at r1, 2 of 4 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75095#-:-MuMA2RR-e7bDqKT4wGM:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

**Pull request merged**